cmd/checkversions/checkversions.sh

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2019 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+#set -x
+
+#debug="--verbose --debug"
+
+runtime_repo="github.com/kata-containers/runtime"
+runtime_repo_dir="$GOPATH/src/${runtime_repo}"
+versions_file="${runtime_repo_dir}/versions.yaml"
+YQ=$(which yq)
+uscan_phrase="uscan-url"
+
+
+debug() {
+	# echo $*
+	true
+}
+
+# Look in the yaml file handed to us to find uscan URLs.
+# return a list of paths to the URL items.
+extract_uscan_items() {
+	vfile=$1
+	local tmpjson=$(mktemp checkverXXX.json)
+	
+	${YQ} r -j "$1" > "$tmpjson"
+	paths=$(jq -r 'paths | join(".")' $tmpjson | grep "${uscan_phrase}$")
+
+	rm $tmpjson
+	echo "$paths"
+}
+
+main() {
+	local fails=0
+	local passes=0
+
+	local paths=$(extract_uscan_items "$versions_file")
+
+	debug "Found [$paths]"
+
+	for path in $paths; do
+		debug ""
+		local rootname="${path%.$uscan_phrase}"
+		echo "Processing [$rootname]"
+		local versionpath="$rootname.version"
+		debug "Version path [$versionpath]"
+		version=$(yq r $versions_file $versionpath)
+		debug "Version [$version]"
+		uscanurl=$(yq r $versions_file $path)
+		debug "uscan URL is [$uscanurl]"
+		local optspath="$rootname.uscan-opts"
+		uscanopts=$(yq r $versions_file $optspath)
+		debug "uscan OPTS is [$uscanopts]"
+
+		debug "Generating watchfile"
+		echo "version=4" > watchfile
+		if [ -n "$uscanopts" ] && [ "$uscanopts" != "null" ]; then
+			echo "$uscanopts  \\" >> watchfile
+		fi
+		echo "$uscanurl" >> watchfile
+		result="$(uscan --report --package ${rootname} --upstream-version ${version} --watchfile watchfile 2>/dev/null)"
+
+		failure=$(grep Newer <<< $result)
+
+		if [ -n "$failure" ]; then
+			echo "----------------------"
+			echo "FAILED: $path"
+			echo "$result"
```

Comment:
I'm not sure how easy this would be, but it could be useful to make the output of this script more machine-readable. You won't be surprised that YAML springs to mind, but even a `--csv` option that dumped out the following fields:

```
package-name;package-uscan-url;current-version;latest-version;pass/fail
```

With that, we could run this as part of the CI and even potentially have a job that attempts to:

- Update `versions.yaml` to the latest version reported by this script.
- Re-run the CI and see what happens.

Reply:
Yeah, I want to make the output nicer, and to do that I think I need to collect things up in a bash table first. joy!  I'm not sure I can easily extract the 'new version' from the uscan though - it doesn't hand it out in a particularly nice way - but, maybe some awk/grep can do that if we are smart. Let me see how nicely I can construct the data in the script, and then we can work on output formats.